File: src-tauri/src/skills.rs
// ──────────────────────────────────────────────────────────
// Phase 4 — Skills
// ──────────────────────────────────────────────────────────
//
// A "skill" is a markdown file with frontmatter living in
// `<project>/skills/`. The chat panel loads the list on mount
// and the user can pre-pend one to any prompt — the skill's
// body becomes the system prompt for that turn.
//
// File format:
//   ---
//   name: Human label
//   description: One-line summary
//   icon: LucideIconName   (optional; defaults to Sparkles)
//   ---
//   <body — becomes the system prompt>
// ──────────────────────────────────────────────────────────

use serde::Serialize;
use std::fs;

use crate::sidecar::SidecarState;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Skill {
    pub id: String,
    pub name: String,
    pub description: String,
    pub icon: String,
    pub body: String,
    pub path: String,
}

/// Parse the `---` frontmatter block. Returns (map, body).
fn split_frontmatter(raw: &str) -> (std::collections::HashMap<String, String>, String) {
    let mut map = std::collections::HashMap::new();
    let trimmed = raw.trim_start();
    if !trimmed.starts_with("---") {
        return (map, raw.to_string());
    }
    // Skip the opening delimiter line.
    let after_open = match trimmed.find('\n') {
        Some(i) => &trimmed[i + 1..],
        None => return (map, raw.to_string()),
    };
    // Find the closing `---`.
    let end = match after_open.find("\n---") {
        Some(i) => i,
        None => return (map, raw.to_string()),
    };
    let fm_text = &after_open[..end];
    for line in fm_text.lines() {
        if let Some((k, v)) = line.split_once(':') {
            map.insert(k.trim().to_string(), v.trim().to_string());
        }
    }
    let body_start = end + "\n---".len();
    let body = after_open[body_start..]
        .trim_start_matches('\n')
        .to_string();
    (map, body)
}

#[tauri::command]
pub fn skills_list(state: tauri::State<'_, SidecarState>) -> Result<Vec<Skill>, String> {
    let root = state
        .current_root()
        .ok_or_else(|| "no project root".to_string())?;
    let skills_dir = root.join("skills");
    if !skills_dir.exists() {
        return Ok(Vec::new());
    }
    let mut out = Vec::new();
    let entries = fs::read_dir(&skills_dir).map_err(|e| format!("read_dir: {}", e))?;
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }
        let id = path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("skill")
            .to_string();
        let raw = match fs::read_to_string(&path) {
            Ok(s) => s,
            Err(_) => continue,
        };
        let (fm, body) = split_frontmatter(&raw);
        out.push(Skill {
            id: id.clone(),
            name: fm.get("name").cloned().unwrap_or(id),
            description: fm.get("description").cloned().unwrap_or_default(),
            icon: fm.get("icon").cloned().unwrap_or_else(|| "Sparkles".into()),
            body,
            path: path.to_string_lossy().into_owned(),
        });
    }
    // Stable ordering by display name so the picker doesn't shuffle
    // between calls.
    out.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));
    Ok(out)
}

File: electron/ipc/commands/skills.ts
// ──────────────────────────────────────────────────────────
// IPC commands: skills — port of src-tauri/src/skills.rs
// ──────────────────────────────────────────────────────────
//
// A "skill" is a markdown file under <project>/skills/ with YAML-
// ish frontmatter. Chat panel loads the list on mount and the user
// can prepend one to any prompt — the body becomes the system
// prompt for that turn.
//
// File format:
//   ---
//   name: Human label
//   description: One-line summary
//   icon: LucideIconName   (optional; defaults to "Sparkles")
//   ---
//   <body — becomes the system prompt>
// ──────────────────────────────────────────────────────────

import fs from "node:fs";
import path from "node:path";
import { currentRoot } from "../../sidecar";
import type { CommandHandler } from "../router";

interface SkillPayload {
  id: string;
  name: string;
  description: string;
  icon: string;
  body: string;
  path: string;
}

/** Split `--- <yaml> ---\n<body>` into a flat key→value map and the
 *  rest of the file. Matches split_frontmatter() in skills.rs: only
 *  top-level `key: value` lines on the yaml side, everything after
 *  the closing `---` (and one leading newline) is body. */
function splitFrontmatter(raw: string): { fm: Map<string, string>; body: string } {
  const fm = new Map<string, string>();
  const trimmed = raw.replace(/^\s+/, "");
  if (!trimmed.startsWith("---")) return { fm, body: raw };

  const firstNewline = trimmed.indexOf("\n");
  if (firstNewline < 0) return { fm, body: raw };
  const afterOpen = trimmed.slice(firstNewline + 1);

  const closingIdx = afterOpen.indexOf("\n---");
  if (closingIdx < 0) return { fm, body: raw };

  const fmText = afterOpen.slice(0, closingIdx);
  for (const line of fmText.split("\n")) {
    const colon = line.indexOf(":");
    if (colon < 0) continue;
    const key = line.slice(0, colon).trim();
    const value = line.slice(colon + 1).trim();
    if (key) fm.set(key, value);
  }

  let body = afterOpen.slice(closingIdx + "\n---".length);
  // Strip a single leading newline after the closing delimiter,
  // matching Rust's trim_start_matches('\n') semantics.
  if (body.startsWith("\n")) body = body.slice(1);
  return { fm, body };
}

export const skillsList: CommandHandler = (args) => {
  void args;
  const root = currentRoot();
  if (!root) throw new Error("no project root");

  const skillsDir = path.join(root, "skills");
  if (!fs.existsSync(skillsDir)) return [];

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(skillsDir, { withFileTypes: true });
  } catch (err) {
    throw new Error(
      `read_dir: ${err instanceof Error ? err.message : String(err)}`,
    );
  }

  const out: SkillPayload[] = [];
  for (const entry of entries) {
    if (entry.isDirectory()) continue;
    if (path.extname(entry.name) !== ".md") continue;

    const id = path.basename(entry.name, ".md");
    const abs = path.join(skillsDir, entry.name);
    let raw: string;
    try {
      raw = fs.readFileSync(abs, "utf-8");
    } catch {
      continue;
    }

    const { fm, body } = splitFrontmatter(raw);
    out.push({
      id,
      name: fm.get("name") ?? id,
      description: fm.get("description") ?? "",
      icon: fm.get("icon") ?? "Sparkles",
      body,
      path: abs,
    });
  }

  // Stable lower-case sort so the picker doesn't shuffle between calls.
  out.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
  return out;
};
